### ISSUE TITLE
Option to have multiple mappings for different actions and not just one.

### ISSUE THREAD

**Issue by rafamadriz** [ruifm/gitlinker.nvim#19]
**Is your feature request related to a problem? Please describe.**

I'd love to have multiple mappings for either opening in browser or just copying to clipboard. Right know you just can have one at the time. This is annoying and frustrating because if I set default callback to be copy to clipboard, then I can't open in browser in case I need or want to (or vice versa) ~~For that, I would need to go into my config, change the default callback, compile packer, reload neovim and finally in can open in browser~~ . Of course I can copy the link to my browser and open it, but it would be super nice to have multiple mappings for that. 

One other feature that would be cool, is just opening the homepage for a repo. I don't know how difficult is this to implement but I guess that most of the url parsing is already done.

**Describe the solution you'd like**

Add options to have multiple mappings for different actions and not just one. 

**Describe alternatives you've considered**

I don't know of other alternatives. 

**Additional context**
 
I tried creating my own mappings by doing something like `lua require"gitlinker.actions".copy_to_clipboard()` but this gave an error and mapping to `open_in_browser` didn't do anything.

**Comment by ruifm** [comment]
> I'd love to have multiple mappings for either opening in browser or just copying to clipboard. Right know you just can have one at the time. This is annoying and frustrating because if I set default callback to be copy to clipboard, then I can't open in browser in case I need or want to (or vice versa) ~For that, I would need to go into my config, change the default callback, compile packer, reload neovim and finally in can open in browser~ . Of course I can copy the link to my browser and open it, but it would be super nice to have multiple mappings for that.

I just merged https://github.com/ruifm/gitlinker.nvim/pull/20 which should make it easier to do what you want.
It's not perfect: you still need to setup extra mappings manually, but at least it's possible.

See the updated https://github.com/ruifm/gitlinker.nvim#usage.
TLDR:

``` lua
vim.api.nvim_set_keymap('n', '<leader>gb', '<cmd>lua require"gitlinker".get_buf_range_url("n", {action_callback = require"gitlinker.actions".open_in_browser})<cr>', {silent = true})
vim.api.nvim_set_keymap('v', '<leader>gb', ':lua require"gitlinker".get_buf_range_url("v", {action_callback require"gitlinker.actions".open_in_browser})<cr>')
```


Until https://github.com/neovim/neovim/pull/13823 is merged in core, I don't think I can do it better since I cannot map to a lua function, I always need to serialize it in a string, which makes passing parameters cumbersome.

Ideally, when that is merged, we would have something like:

```lua
require"gitlinker".setup({
  mappings = {
    "<leader>gy" = require"gitlinker".actions.copy_to_clipboard,
    "<leader>gb" = require"gitlinker".actions.open_in_browser,
  }
})
```
but unfortunately, I really can't do that right now since I cannot serialize a lua function to vimscript (without making it public, which is terrible).

A workaround would be to stop accepting arbitrary funcions as `action_callback`s, and just have a limited set of possible functions that can be identified with a string ID. I don't like the idea at all, it goes against the point of lua in terms of extensibility.

> 
> One other feature that would be cool, is just opening the homepage for a repo. I don't know how difficult is this to implement but I guess that most of the url parsing is already done.

Sure, sounds simple enough. You can probably do something like that already with user configuration already, but I can make it easier by passing `url_data` as second parameter to the `action_callback` function.

I can open a PR for that later this week.

> I tried creating my own mappings by doing something like `lua require"gitlinker.actions".copy_to_clipboard()` but this gave an error and mapping to `open_in_browser` didn't do anything.

That is because those are callback functions, i.e., they are supposed to be called by `require"gitlinker".get_buf_range_url()`, and that is the function you should be mapping to.
Functions in `gitlinker/actions` are just callbacks that receive the generated `url` as a paramter and do an arbitrary thing with it. You can even write your own crazy `action_callback` for instance, send an email/IM to someone with the url.

**Comment by ruifm** [comment]
Also, let me know if you consider this resolved, we can move the "open repo homepage" in a new issue.

**Comment by rafamadriz** [comment]
> Until neovim/neovim#13823 is merged in core, I don't think I can do it better since I cannot map to a lua function, I always need to serialize it in a string, which makes passing parameters cumbersome.

Nah, don't worry. I understand, and all my mappings look like that, just a string in the place where should be a command/function. Looks ugly but hopefully #138223 gets merged soon.

And thanks for adding this.

**Comment by rafamadriz** [comment]
neovim/neovim#16591 got merged (which supersedes neovim/neovim#13823). 

Is it possible to now properly implement having multiple mappings for either opening in browser or just copying to clipboard ?
